### DISCUSSION TITLE
Does PipeCD conform to GitOps Principles?

### DISCUSSION THREAD

**Discussion by lloydchang** [pipe-cd/pipecd#2752] [Q&A] [answered]
Hi, @nghialv,

Thank you for your time.

❔ Does PipeCD conform to [GitOps Principles](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md)?

❔ What are the definitions of the hashtag `#gitops` and the word "GitOps" from PipeCD's perspective?

❔ Perhaps PipeCD defines the word "GitOps" differently from panelists in [Panel Discussion: OpenGitOps and the GitOps Working Group](https://www.youtube.com/watch?v=p0ntvMyqmr4)?

Specifically, my concern about PipeCD's usage of the word "GitOps" is:
> ["People have also been doing things that don't follow the GitOps Principles and because they are using Git, they believe they're doing GitOps."](https://www.youtube.com/watch?v=p0ntvMyqmr4&t=528s)

_Please correct me if my understanding is incorrect._

My understanding is:

* While PipesCD uses the hashtag `#gitops` and the word "GitOps", PipeCD executes ["CIOps" _push-based_ deployments](https://www.weave.works/blog/kubernetes-anti-patterns-let-s-do-gitops-not-ciops) via [PipeCD's triggering mechanism](https://pipecd.dev/docs/user-guide/triggering-a-deployment/), instead of [Pull Automatically in GitOps Principles](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md).

* While [PipeCD's configuration drift detection](https://pipecd.dev/docs-dev/user-guide/configuration-drift-detection/) exists, that is different from [Continuously Reconciled in GitOps Principles](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md).

* Since [`piped` is a stateless component](https://cloud-native.slack.com/archives/C01B27F9T0X/p1602479098014800) and [released as a single binary for Kubernetes, Terraform, CloudRun, Lambda, ECS](https://cloud-native.slack.com/archives/C01B27F9T0X/p1602479256017500), PipeCD doesn't use [Kubernetes control loops](https://kubernetes.io/docs/concepts/architecture/controller/).

Thank you for answering my questions.

**Comment by nghialv** [comment, marked as answer]
Hi, @lloydchang 

Firstly, thank you for your question. ❤️ 

> Does PipeCD conform to GitOps Principles?

Yes, I think that PipeCD conforms to those GitOps Principles.

>  What are the definitions of the hashtag #gitops and the word "GitOps" from PipeCD's perspective?

- All system resources should be defined declaratively and stored immutably (only adding commit, no forced-push) in Git. 
- All operations should be done via Git operations.
- The CD system will pull the desired state defined in Git and attempt to apply them. (A deployment is a process that does the transition from the current state (running state) to the desired state (the specified state in Git) of a specific application.)

> Perhaps PipeCD defines the word "GitOps" differently from panelists in Panel Discussion: OpenGitOps and the GitOps Working Group?

I don't think there's that much of a difference. Can you tell me the difference you are thinking of?

---
Please let me correct something about PipeCD in your comments.

> While PipesCD uses the hashtag #gitops and the word "GitOps", PipeCD executes "CIOps" push-based deployments via PipeCD's triggering mechanism, instead of Pull Automatically in GitOps Principles.

No, PipeCD is not doing CIOps like that. As you know PipeCD has 2 components: control-plane and agent (piped).
The agent can be installed inside any cluster, any machine and it will periodically pull the desired state from the defined Git repositories to trigger the needed deployments to make the applications close to the desired state. The SYNC button on the web console is just a way for those who want to manually trigger new sync from the web.

![image](https://user-images.githubusercontent.com/1751755/140749352-3c5dece0-f1dd-44c4-bf75-f24cc89ae305.png)

> While PipeCD's configuration drift detection exists, that is different from Continuously Reconciled in GitOps Principles.

By default, PipeCD continuously detects the drift to report/alert to users (by showing the diff on the web console, sending notification to Slack...). Currently, PipeCD does not re-apply (re-trigger the deployment) when detected a drift but we're on the way to support that. And actually, not all users want to enable that auto re-applying feature. So PipeCD will provide a way to allow users to customize the trigger logic to fit with their situation. It will be something like:

``` yaml
apiVersion: pipecd.dev/v1beta1
kind: Terraform
spec:
  ...
  trigger:
    # trigger a new deployment when any resource is out-of-sync.
    onOutOfSync:
      disable: true   # default is `true`
    # trigger a new deployment when received a command from web-console, pipectl or api
    onCommand:
      disable: false  # default is `false`
    # trigger a new deployment when the newly merged commits touch any resources of an application.
    onCommit:
      disable: false  # default is `false`
```

> Since piped is a stateless component and released as a single binary for Kubernetes, Terraform, CloudRun, Lambda, ECS, PipeCD doesn't use Kubernetes control loops.

Yes, Piped agent is a stateless component. A Piped agent can be installed in the Kubernetes cluster as a single pod (to deploy Kubernetes applications running inside that cluster), run as an AWS Fargate service, run as a CloudRun service, or run inside any machine. ([Installation guides](https://pipecd.dev/docs/operator-manual/piped/installation/))
Via Piped agent, PipeCD can deploy Kubernetes application as well as Terraform, CloudRun, Lambda, ECS applications.
Yes, PipeCD is not only for K8s so Piped agent does not use Kubernetes control loops. But it is having something similar to that inside.
Btw, I think GitOps is not just for K8s so "using Kubernetes control loop or not" should not be the thing to be considered.

Thanks.
Please let me know if you have any questions.

**Reply by lloydchang** [reply]
Thank you @nghialv — Related, please see [#wg-gitops Slack channel](https://cloud-native.slack.com/archives/C01G9DEE88M/p1636387969104100?thread_ts=1635251844.014900) at https://cloud-native.slack.com/archives/C01G9DEE88M/p1636387969104100?thread_ts=1635251844.014900

**Reply by nghialv** [reply]
Thank you. Let me check that slack channel.

**Reply by lloydchang** [reply]
@nghialv asked:
> Can you tell me the difference you are thinking of?

I empathize with you that not all users of PipeCD want to enable that auto re-applying feature.

However... Aspirationally, [_attempt to apply_](https://github.com/open-gitops/documents/blob/v1.0.0/GLOSSARY.md#reconciliation) is a _requirement_ in [GitOps Principle 4](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md):
> 4. **Continuously Reconciled**
>
>    Software agents [continuously](./GLOSSARY.md#continuous) observe actual system state and [attempt to apply](./GLOSSARY.md#reconciliation) the desired state.

Details at https://cloud-native.slack.com/archives/C01G9DEE88M/p1636487514105300?thread_ts=1635251844.014900

@nghialv @christianh814 Thank you.

**Reply by nghialv** [reply]
Hi @lloydchang 

We have just released a new version v0.22.0, which will allow Piped agent to attempt to apply the desired state as you said.

https://github.com/pipe-cd/pipe/releases/tag/v0.22.0
